CRM/Core/Payment.php

```diff
@@ -1807,4 +1807,59 @@ public function isSendReceiptForPending() {
     return FALSE;
   }
 
+  /**
+   * Update recurring contribution based on incoming payment.
+   *
+   * @param int $recurringContributionID
+   * @param string $paymentStatus
+   *   Payment status - this correlates to the machine name of the contribution status ID ie
+   *   - Completed
+   *   - Failed
+   * @param string $effectiveDate
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function updateOnNewPayment($recurringContributionID, $paymentStatus, $effectiveDate) {
+
+    $effectiveDate = $effectiveDate ? date('Y-m-d', strtotime($effectiveDate)) : date('Y-m-d');
+    if (!in_array($paymentStatus, ['Completed', 'Failed'])) {
+      return;
+    }
+    $params = [
+      'id' => $recurringContributionID,
+      'return' => [
+        'contribution_status_id',
+        'next_sched_contribution_date',
+        'frequency_unit',
+        'frequency_interval',
+        'installments',
+        'failure_count',
+      ],
+    ];
+
+    $existing = civicrm_api3('ContributionRecur', 'getsingle', $params);
+
+    if ($paymentStatus == 'Completed'
+      && CRM_Contribute_PseudoConstant::contributionStatus($existing['contribution_status_id'], 'name') == 'Pending') {
+      $params['contribution_status_id'] = 'In Progress';
+    }
+    if ($paymentStatus == 'Failed') {
+      $params['failure_count'] = $existing['failure_count'];
+    }
+    $params['modified_date'] = date('Y-m-d H:i:s');
+
+    if (!empty($existing['installments']) && self::isComplete($recurringContributionID, $existing['installments'])) {
+      $params['contribution_status_id'] = 'Completed';
+    }
+    else {
+      // Only update next sched date if it's empty or 'just now' because payment processors may be managing
+      // the scheduled date themselves as core did not previously provide any help.
+      if (empty($existing['next_sched_contribution_date']) || strtotime($existing['next_sched_contribution_date']) ==
+        strtotime($effectiveDate)) {
+        $params['next_sched_contribution_date'] = date('Y-m-d', strtotime('+' . $existing['frequency_interval'] . ' ' . $existing['frequency_unit'], strtotime($effectiveDate)));
+      }
```

Comment:
Is your idea to remove the logic here guarding the change to next_sched_contribution_date so that it is always updated if that information hasn't been provided?

Reply:
@artfulrobot @andrew-cormick-dockery I'd like to keep the logic changes out of this PR and just focus on the implementation so we can get an overridable method into core that payment processors can start using. It's clear that there's quite a few "fixes"/"changes" that need to be made to the core logic.

Reply:
I'll leave it to you all for strategy in terms of how to solve this problem.  I acknowledge that I didn't personally have an easy way forward.  What I'll do is advocate for our particular use case to ensure that it's resolved in the end.

(In other words, go for it).